trino/client/client.go

```diff
@@ -0,0 +1,260 @@
+	}
+
+	if c.User != "" && c.Password == "" {
+		return "", fmt.Errorf("password must be provided if user is set")
+	}
+
+	if c.User != "" && c.Password != "" {
+		if !encrypted {
+			return "", fmt.Errorf("encryption must be abled for user authentication")
+		}
+		rawDSN = fmt.Sprintf("https://%s:%s@", c.User, c.Password)
+	}
+
+	rawDSN += c.ServerHost
+
+	return rawDSN, nil
+}
+
+// Client is a Trino client that provides methods to connect and query a Trino server
+// it also provides exponential backoff retry logic for queries.
+type Client struct {
+	// config is the trino.Config used to generate a DSN for the connection.
+	config trino.Config
+	// dsn is the Data Source Name used to connect to the Trino server.
+	dsn string
+	// conn is the SQL database connection to the Trino server.
+	conn *sql.DB
+	// retryCount is the number of times to retry a query in case of failure, default 5
+	retryCount int
+}
+
+func New(uri string, opts ...option) (*Client, error) {
+	c := &Client{
+		config: trino.Config{
+			ServerURI: uri,
+		},
+		retryCount: 5, // default retry count
+	}
+
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
+		}
+	}
+
+	dsn, err := c.config.FormatDSN()
+	if err != nil {
+		// cannot log error directly because DSN contains username and password
+		return nil, fmt.Errorf("malformed server URI, please check your connection string and try again")
+	}
+	c.dsn = dsn
+
+	return c, nil
+}
+
+func (c *Client) Connect() error {
+	db, err := sql.Open("trino", c.dsn)
+	if err != nil {
+		if db != nil {
+			db.Close() //nolint:errcheck
+		}
+		// cannot log error directly because DSN contains username and password
+		return fmt.Errorf("malformed server URI, please check your connection string and try again")
+	}
+	c.conn = db
+	log.Info("Connected to Trino")
+
+	return nil
+}
+
+// Disconnect closes the connection to the Trino server.
+// it is intended to be used as a defer function so no error is returned.
+func (c *Client) Disconnect() {
+	c.conn.Close() //nolint:errcheck
+	log.Info("Connection to Trino closed")
+}
+
+func (c *Client) Query(ctx context.Context, statement string, args ...any) (*sql.Rows, error) {
+	if c.conn == nil {
+		if e := c.Connect(); e != nil {
+			return nil, e
+		}
+	}
+
+	// Trino conn.PrepareContext doesn't actually connect to the DB it just returns a *Stmt
+	stmt, err := c.conn.PrepareContext(ctx, statement)
+	if err != nil {
+		// this is unreavchable for the trino driver but we handle it to keep the linter happy
+		return nil, fmt.Errorf("prepare Error: %s", err)
+	}
+	// this cannot return an error for trino
+	defer stmt.Close() //nolint:errcheck
+
+	count := 0
+	for {
+		out, err := stmt.QueryContext(ctx, args...)
+		if err == nil {
+			return out, nil
+		}
+
+		count++
+		if count > c.retryCount {
+			return nil, fmt.Errorf("query failed after %d retries: %w", count, err)
+		}
+
+		time.Sleep(time.Second * 2 * time.Duration(count)) // exponential backoff
```

Comment:
nit: should the starting 2 seconds be configurable as well?

Reply:
too many knobs.